src/doorman/test/TelnetProtocolTest.cpp

```diff
@@ -0,0 +1,270 @@
+#include "TelnetProtocol.hpp"
+
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include <arpa/telnet.h>
+#include <catch2/catch.hpp>
+#include <catch2/trompeloeil.hpp>
+
+namespace {
+class MockHandler : public TelnetProtocol::Handler {
+public:
+    MAKE_MOCK1(send_bytes, void(gsl::span<const byte>), override);
+    MAKE_MOCK1(on_line, void(std::string_view), override);
+    MAKE_MOCK2(on_terminal_size, void(unsigned int, unsigned int), override);
+    MAKE_MOCK2(on_terminal_type, void(std::string_view type, bool ansi_supported), override);
+};
+
+using bytes = std::vector<byte>;
+auto wrap(std::string_view sv) {
+    bytes b;
+    for (auto c : sv)
+        b.emplace_back(static_cast<byte>(c));
+    return b;
+}
+auto match_bytes(const bytes &bytes) {
+    return trompeloeil::make_matcher<gsl::span<const byte>>(
```

Comment:
It's interesting how similar this is to Java's Hamcrest Matcher. Same concept, just C++ style.